feat(connector-polkadot): add connector pkg, openapi specs, test suite

AnmolBansalDEV opened the pull request:
It's the updated PR for Polkadot support for Cacti that incorporates the efforts of @RafaelAPB  @CatarinaPedreira along with my work during the Hyperledger Mentorship under the mentorship of @jagpreetsinghsasan @suvajit-sarkar @petermetz 

## COMMITS TO REVIEW
feat(polkadot): creation of polkadot AIO image
```
Primary Changes
---------------
1. Updated docker image to include multi-stage build
2. Added healthcheck to the image
3. Added Supervisord and removed unneccessary start script
4. Updated the Readme with relevant commands
```
feat(polkadot): update substrate test tooling
```
Primary Changes
---------------
1. Added correct healthcheck for ledger container
2. Update the Substrate test ledger testcases
```
feat(polkadot): creation of polkadot connector openapi specs, test case and connector class
```
Primary Changes
---------------
1. Created openapi specs for get-prometheus-exporter-metrics, get-transaction-info,
   get-raw-transaction, sign-raw-transaction, run-transaction, deploy-contract-ink,
   and invoke-contract endpoints
2. Created relevant types for the request as well as response for the above endpoints
3. Added generated code for these endpoints
4. Created connector class with functions to interact with the polkadot ledger
5. Created webservices to interact with the endpoint and the relevant class method
6. Created unit and integration testcases in jest to test base functionality of the connector

Secondary Changes
-----------------
1. Added an ink! contract for running the testcases
2. Added the polkadot connector to ci workflow
3. Created substrate.md to docs-cactus
4. Added the polkadot connector to tsconfig.json
```
feat(polkadot): creation of readme and architecture reference diagrams
```
Primary Changes
---------------
1. Added README.md for the connector
2. Added Architecture diagrams for the plugin

Secondary Changes
-----------------
1. Added the docker image for the polkadot connector
```




**Pull Request Requirements**
- [x] Rebased onto `upstream/main` branch and squashed into single commit to help maintainers review it more efficient and to avoid spaghetti git commit graphs that obfuscate which commit did exactly what change, when and, why.
- [x] Have git sign off at the end of commit message to avoid being marked red. You can add `-s` flag when using `git commit` command. You may refer to this [link](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) for more information.
- [x] Follow the Commit Linting specification. You may refer to this [link](https://www.conventionalcommits.org/en/v1.0.0-beta.4/#specification) for more information. 

**Character Limit**
- [x] Pull Request Title and Commit Subject must not exceed 72 characters (including spaces and special characters).
- [x] Commit Message per line must not exceed 80 characters (including spaces and special characters).

**A Must Read for Beginners**
For rebasing and squashing, here's a [must read guide](https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing) for beginners.

jagpreetsinghsasan commented:
@AnmolBansalDEV can you squash merge commits into the logical ones?

jagpreetsinghsasan commented on packages/cactus-plugin-ledger-connector-polkadot/README.md:
```diff
@@ -0,0 +1,67 @@
+# `@hyperledger/cactus-plugin-ledger-connector-polkadot`
```
Please refer to other plugins readme for reference (Example: https://github.com/hyperledger/cacti/blob/main/packages/cactus-plugin-ledger-connector-fabric/README.md)

jagpreetsinghsasan replied:
Also, you can add the referenced UML diagrams here itself so that I can review it all together

AnmolBansalDEV replied:
done

jagpreetsinghsasan commented on package.json:
```diff
@@ -53,8 +53,9 @@
     "codegen:warmup-mkdir": "make-dir ./node_modules/@openapitools/openapi-generator-cli/versions/",
-    "codegen:warmup-v6.6.0": "nwget https://repo1.maven.org/maven2/org/openapitools/openapi-generator-cli/6.6.0/openapi-generator-cli-6.6.0.jar -O ./node_modules/@openapitools/openapi-generator-cli/versions/6.6.0.jar",
-    "watch-other": "lerna run --parallel watch",
+    "codegen:warmup-v5.2.1": "nwget https://repo1.maven.org/maven2/org/openapitools/openapi-generator-cli/5.2.1/openapi-generator-cli-5.2.1.jar -O ./node_modules/@openapitools/openapi-generator-cli/versions/5.2.1.jar",
```
The removed changes were the correct ones

AnmolBansalDEV replied:
reverted.

jagpreetsinghsasan commented on packages/cactus-plugin-ledger-connector-polkadot/docs/architechture/run-transaction-endpoint.puml:
You can add the UML diagrams

AnmolBansalDEV replied:
added

jagpreetsinghsasan commented on packages/cactus-plugin-ledger-connector-polkadot/package.json:
```diff
@@ -0,0 +1,114 @@
+{
+  "name": "@hyperledger/cactus-plugin-ledger-connector-polkadot",
+  "version": "1.1.2",
```
Please update it to the latest version `2.0.0-alpha.2`

AnmolBansalDEV replied:
changed

jagpreetsinghsasan commented on packages/cactus-plugin-ledger-connector-polkadot/package.json:
```diff
@@ -0,0 +1,114 @@
+  },
+  "homepage": "https://github.com/hyperledger/cactus#readme",
+  "dependencies": {
+    "@hyperledger/cactus-common": "1.1.2",
```
Please update all the dependencies to the latest version `2.0.0-alpha.2` : [Ref](https://github.com/hyperledger/cacti/blob/8addb018b6d124d54d9d948bbaeba6ea33b67153/packages/cactus-plugin-ledger-connector-besu/package.json#L57)

AnmolBansalDEV replied:
done

jagpreetsinghsasan commented on packages/cactus-plugin-ledger-connector-polkadot/package.json:
```diff
@@ -0,0 +1,114 @@
+    "uuid": "8.3.2"
+  },
+  "devDependencies": {
+    "@hyperledger/cactus-plugin-keychain-memory": "2.0.0-alpha.1",
```
Same here as above

AnmolBansalDEV replied:
same as above

jagpreetsinghsasan commented on packages/cactus-plugin-ledger-connector-polkadot/src/main/typescript/plugin-ledger-connector-polkadot.ts:
```diff
@@ -0,0 +1,901 @@
+  public setProvider(wsProviderUrl: string): void {
+    try {
+      this.wsProvider = new WsProvider(wsProviderUrl, this.autoConnect);
+    } catch (e) {
```
Please give this var a meaningful name `err` (all the places)

AnmolBansalDEV replied:
implemented

jagpreetsinghsasan commented on packages/cactus-plugin-ledger-connector-polkadot/src/main/typescript/plugin-ledger-connector-polkadot.ts:
```diff
@@ -0,0 +1,901 @@
+    try {
+      this.api = await ApiPromise.create({ provider: this.wsProvider });
+    } catch (e) {
+      throw Error("Could not create API");
```
Please print the error `err` as well (after renaming the var)

AnmolBansalDEV replied:
👍

jagpreetsinghsasan commented:
Overall, great work done @AnmolBansalDEV :sunflower:

outSH commented on packages/cactus-plugin-ledger-connector-polkadot/src/main/typescript/plugin-ledger-connector-polkadot.ts:
```diff
@@ -0,0 +1,901 @@
+      success = true;
+      return { success, callOutput };
+    } else if (req.invocationType === PolkadotContractInvocationType.Send) {
+      if (isWeb3SigningCredentialNone(req.web3SigningCredential)) {
+        throw new Error(`${fnTag} Cannot invoke contract with pre-signed TX`);
+      }
+      let mnemonic: string;
+      if (isWeb3SigningCredentialMnemonicString(req.web3SigningCredential)) {
+        const web3SigningCredential =
+          req.web3SigningCredential as Web3SigningCredentialMnemonicString;
+        mnemonic = web3SigningCredential.mnemonic;
+      } else if (isWeb3SigningCredentialCactusRef(req.web3SigningCredential)) {
+        const web3SigningCredential =
+          req.web3SigningCredential as Web3SigningCredentialCactusKeychainRef;
+        const { keychainEntryKey, keychainId } = web3SigningCredential;
+        // locate the keychain plugin that has access to the keychain backend
+        // denoted by the keychainID from the request.
+        const keychainPlugin =
+          this.pluginRegistry.findOneByKeychainId(keychainId);
+
+        Checks.truthy(
+          keychainPlugin,
+          `${fnTag} keychain for ID:"${keychainId}"`,
+        );
+
+        // Now use the found keychain plugin to actually perform the lookup of
+        // the private key that we need to run the transaction.
+        mnemonic = await keychainPlugin?.get(keychainEntryKey);
+      } else {
+        throw new Error(
+          `${fnTag} Unrecognized Web3SigningCredentialType: ` +
+            `Supported ones are: ` +
+            `${Object.values(Web3SigningCredentialType).join(";")}`,
+        );
+      }
```
Common code like this can be extracted to a separate (private) method (both here and in `deployContract`)

AnmolBansalDEV replied:
done

outSH commented on packages/cactus-plugin-ledger-connector-polkadot/src/main/typescript/web-services/deploy-contract-ink-endpoint.ts:
```diff
@@ -0,0 +1,99 @@
+      this.log.error(`Crash while serving ${reqTag}`, ex);
+      res.status(500).json({
+        message: "Internal Server Error",
+        error: ex?.stack || ex?.message,
```
Please use `safeStringifyException` (https://github.com/hyperledger/cacti/blob/main/packages/cactus-common/src/main/typescript/error-utils.ts#L29) to sanitize output and handle cases when exception is not an instance of `Error`. Same applies to all other endpoints as well.

AnmolBansalDEV replied:
done

outSH commented on packages/cactus-plugin-ledger-connector-polkadot/src/main/json/openapi.json:
```diff
@@ -0,0 +1,781 @@
+        "summary": "Get the Prometheus Metrics",
+        "parameters": [],
+        "responses": {
+          "200": {
```
Please add description of error responses as well (similar to this - https://github.com/hyperledger/cacti/blob/main/packages/cactus-plugin-ledger-connector-ethereum/src/main/json/openapi.json#L1150). This applies to other endpoints as well

AnmolBansalDEV replied:
done

outSH commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/api-surface.test.ts:
```diff
@@ -0,0 +1,8 @@
+import test, { Test } from "tape-promise/tape";
```
Tape tests must be ignored in `jest.config.js` and listed in `.taprc`.

Also, it would be nice to use jest instead of tape (ignore it if it's too much effort, we can migrate it some other time).

AnmolBansalDEV replied:
migrated to jest

outSH commented on packages/cactus-plugin-ledger-connector-polkadot/package.json:
```diff
@@ -0,0 +1,114 @@
+{
+  "name": "@hyperledger/cactus-plugin-ledger-connector-polkadot",
```
- Add this package to tsconfig references (https://github.com/hyperledger/cacti/blob/main/tsconfig.json#L3) to include it in build
- Add this package to CI (like that - https://github.com/hyperledger/cacti/blob/main/.github/workflows/ci.yaml#L200, but set JEST_TEST_RUNNER_DISABLED to false if this package use tape tests only.

AnmolBansalDEV replied:
added the package to tsconfig, second not needed as tests are now in jest

outSH replied:
@AnmolBansalDEV CI entry (P2) is still needed, but with `JEST_TEST_RUNNER_DISABLED` set to false and `TAPE_TEST_RUNNER_DISABLED` set to `true`, something like the following:

```
  cactus-plugin-ledger-connector-polkadot:
    continue-on-error: false
    env:
      FULL_BUILD_DISABLED: true
      JEST_TEST_PATTERN: packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/(unit|integration|benchmark)/.*/*.test.ts
      JEST_TEST_RUNNER_DISABLED: false
      TAPE_TEST_RUNNER_DISABLED: true
    needs: build-dev
    runs-on: ubuntu-20.04
    steps:
      - name: Use Node.js ${{ env.NODEJS_VERSION }}
        uses: actions/setup-node@v3.6.0
        with:
          node-version: ${{ env.NODEJS_VERSION }}
      - uses: actions/checkout@v3.5.2

      - id: yarn-cache
        name: Restore Yarn Cache
        uses: actions/cache@v3.3.1
        with:
          key: ${{ runner.os }}-yarn-${{ hashFiles('./yarn.lock') }}
          path: ./.yarn/
          restore-keys: |
            ${{ runner.os }}-yarn-${{ hashFiles('./yarn.lock') }}
      - run: ./tools/ci.sh
```

outSH replied:
Also, when you're done with requested changes please hit `re-request review` circle button, next to reviewer avatar in top-right pane to notify us that there's something to see here :)

AnmolBansalDEV replied:
sure, updated. And requested your kind review.

outSH reviewed (changes requested):
Great work, thank you!

I've added some things to fix before merge.

AnmolBansalDEV commented on packages/cactus-plugin-ledger-connector-polkadot/src/main/json/openapi.json:
```diff
@@ -0,0 +1,781 @@
+  "info": {
+    "title": "Hyperledger Cactus Plugin - Connector Polkadot",
+    "description": "Can perform basic tasks on a Polkadot parachain",
+    "version": "0.0.1",
```
updated.

petermetz reviewed (changes requested):
@AnmolBansalDEV 

1. Exception throws and re-throws are done with the appropriate utilities which are either 
* `createRuntimeErrorWithCause()` from the file at `packages/cactus-common/src/main/typescript/exception/create-runtime-error-with-cause.ts` or
* The `http-errors` library's function collection accessed via `createHttpError` .  For HTTP REST API endpoint related error handling where your exceptions communicate with the web layer, have a look at the patterns established in this diff 
https://github.com/hyperledger/cacti/pull/2869
<br>In short, instead of this:
  ```typescript
  // This is wrong - the HTTP response's status code will be
  // set to 500 but it should be 400 because it was user error
  // not a bug in the server's production code.
  throw new Error(
  `${fnTag} Cannot deploy contract without keychainId and the contractName`
	);
	```
	Do this:
	```typescript
	if (!keychainHasContract) {
	// This is better, because the HTTP response will 
	// contain a status code of 400 instead of 500
	const errorMessage =
		`${fnTag} Cannot create an instance of the contract instance because` +
		`the contractName in the request does not exist on the keychain`;
	throw new createHttpError[400](errorMessage);
	}
	```
2. Ensure that your are not swallowing valuable stack traces / exception messages by incorrectly serializing inner exceptions such as shown below (which is only one of many creative ways people accidentally come up with to destroy crash information). The below code will print `Something bad happened instead of something cool: [Object object]`
	```typescript
	try {
		await doSomethingCool();
	}
	catch (ex: unknown) {
		const errorMessage = `Something bad happened instead of something cool: ${ex}`;
		throw new RuntimeError(errorMessage);
	}
	``` 
2. Ensure Linear git history by squashing your commits into a single one and then rebasing onto upstream/main (after having ran a `git fetch --all`)
3. Ensure that `yarn run custom-checks` is not failing with issues in the code that you are modifying/adding.
4. Ensure that the code you have written has test coverage in general and also, more specifically that negative cases are verified by your tests as well. An example for the latter through test cases that send in incorrect input intentionally and then verify that the exception thrown accurately explains what went wrong. An example of this principlan implemented in action can be seen at the test case here: `packages/cactus-cmd-api-server/src/test/typescript/integration/jwt-unprotected-endpoint-authz-ops-confirm.test.ts`.
5. Ensure that your yarn.lock file is up to date. If you haven't added/modified/removed any npm dependencies then this step is not necessary. Otherwise you can ensure an up to date yarn.lock file by running `yarn intsall`. It is important to **NOT** delete the lock file and then recreate it from scratch because this might trigger unintended auto-upgrades on all transitive dependencies that are set to auto-upgrade by the maintainers of the packages that we depend on.
6. Ensure that you are not using container images that you've self-published when it comes to testing infrastructure. If you've made changes to a container image (usually by altering a `Dockerfile`) then you can ask a maintainer to publish those changes to you to the official GitHub Container Registry repository of the project (e.g. ghcr.io/hyperledger/cacti-some-image-that-you-are-working-on)
7. Ensure that your commit message describes what is being changed and if there are multiple things then please try to break it down into multiple pull requests for easier review. Studies show that the larger the pull requests of a repository are on average, the worse code quality usually is.

AnmolBansalDEV commented:
Thank you @petermetz for your review, I've incorporated the changes and I request your re-review.

1. Updated the Errors thrown and have used the `create-errors` lib
2. Parsed the error traces appropriately inside of webservices like this one https://github.com/hyperledger/cacti/blob/a24d90c7a5a41e40772bb7ecef2acf66f87ad7f7/packages/cactus-plugin-ledger-connector-polkadot/src/main/typescript/web-services/sign-raw-transaction-endpoint.ts#L95-L98
3. Cleaned git history by rebasing and having a progressive commit history
4. `yarn run custom-checks` is not failing due to my changes
5. testcases cover negative cases as well like here https://github.com/hyperledger/cacti/blob/a24d90c7a5a41e40772bb7ecef2acf66f87ad7f7/packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts#L128-L143
6. `yarn.lock` contains changes for polkadot connector packages only
7.  Please publish the updated AIO image for the Substrate network present here https://github.com/hyperledger/cacti/tree/a24d90c7a5a41e40772bb7ecef2acf66f87ad7f7/tools/docker/substrate-all-in-one
 Also Kindly publish this polkadot connector image as well https://github.com/hyperledger/cacti/blob/0a57c36f287f18c603eb4acfb68d40f845381b8f/packages/cactus-plugin-ledger-connector-polkadot/dockerfile
8. I've tried to keep the git history streamlined, please suggest changes if you don't find the history fit

petermetz commented on packages/cactus-plugin-ledger-connector-polkadot/README.md:
````diff
@@ -0,0 +1,200 @@
+In the Cactus project root say:
+
+```sh
+DOCKER_BUILDKIT=1 docker build -f ./packages/cactus-plugin-ledger-connector-polkadot/Dockerfile . -t cplcb
````
@AnmolBansalDEV I just tried to run this and it failed with the log pasted below in my comment.

One thing I noticed that you are still installing the old yarn v1 but we've migrated to v3 in the meantime and that ships with NodeJS out of the box too, IIRC.
What I recommend is taking a look at the `Dockerfile` of `cmd-api-server` and taking a cue from over there (you can also use the newer versions of that as a base image)

```sh
$ DOCKER_BUILDKIT=1 docker build -f ./packages/cactus-plugin-ledger-connector-polkadot/Dockerfile . -t cplcb
[+] Building 1.5s (6/6) FINISHED                                                                                                                                                                                                                  
 => [internal] load build definition from Dockerfile                                                                                                                                                                                         0.0s
 => => transferring dockerfile: 309B                                                                                                                                                                                                         0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                            0.0s
 => => transferring context: 887B                                                                                                                                                                                                            0.0s
 => [internal] load metadata for ghcr.io/hyperledger/cactus-cmd-api-server:2022-08-05-7309f2a                                                                                                                                                0.5s
 => [1/3] FROM ghcr.io/hyperledger/cactus-cmd-api-server:2022-08-05-7309f2a@sha256:059d1cc7312ff4a9d10361003176b1be8080a4e0c992c4863c28caf27db60465                                                                                          0.0s
 => CACHED [2/3] RUN npm install -g yarn@1.22.17                                                                                                                                                                                             0.0s
 => ERROR [3/3] RUN yarn add @hyperledger/cactus-plugin-ledger-connector-polkadot@latest --production --ignore-engines                                                                                                                       0.9s
------                                                                                                                                                                                                                                            
 > [3/3] RUN yarn add @hyperledger/cactus-plugin-ledger-connector-polkadot@latest --production --ignore-engines:                                                                                                                                  
#5 0.452 yarn add v1.22.17                                                                                                                                                                                                                        
#5 0.462 warning package.json: No license field                                                                                                                                                                                                   
#5 0.516 warning No license field
#5 0.518 [1/4] Resolving packages...
#5 0.807 error An unexpected error occurred: "https://registry.yarnpkg.com/@hyperledger%2fcactus-plugin-ledger-connector-polkadot: Not found".
#5 0.807 info If you think this is a bug, please open a bug report with the information provided in "/usr/src/app/yarn-error.log".
#5 0.807 info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
------
executor failed running [/bin/bash -c yarn add @hyperledger/cactus-plugin-ledger-connector-polkadot@${NPM_PKG_VERSION} --production --ignore-engines]: exit code: 1
```

AnmolBansalDEV replied:
updated the name of the file, `cmd-api-server` package version, also the yarn version.
https://github.com/hyperledger/cacti/blob/87ec994040fade0e7a4a690883c971baa9eb9702/packages/cactus-plugin-ledger-connector-polkadot/Dockerfile#L1-L7

I installed the node v20 in a Ubuntu container, but it didn't come with `yarn` pre-installed.

>#5 0.807 error An unexpected error occurred: "https://registry.yarnpkg.com/@hyperledger%2fcactus-plugin-ledger-connector-polkadot: Not found".
the log mentions that the `@hyperledger/cactus-plugin-ledger-connector-polkadot` package is not found, which is true as it is not published to NPM registry. Please publish the package.

AnmolBansalDEV replied:
@petermetz I've updated the Dockerfile to remove some issues while building it. It's now building successfully with the published canary version `@hyperledger/cactus-plugin-ledger-connector-polkadot@2.0.0-lfx-connector.163` but it runs into an error during startup. I think it's due to the absence of `onPluginInit()` method cause the `cactus-cmd-api-server` expects it. With the release of a version with new changes, I think it should work fine. I've updated the README.md regarding running the container as well.

https://github.com/hyperledger/cacti/blob/f7e7a2a1cd9694327e4d323650bd3fad9b3b80dd/packages/cactus-plugin-ledger-connector-polkadot/Dockerfile#L1-L10

petermetz replied:
@AnmolBansalDEV Gotcha, that makes sense. I've done a lot of publishing just now.
npm package: `@hyperledger/cactus-plugin-ledger-connector-polkadot@2.0.0-lfx-connector.164`
connector container: `ghcr.io/hyperledger/cactus-connector-polkadot:2024-01-16--2.0.0-lfx-connector.164--pr-2877`
ledger container: `ghcr.io/hyperledger/cactus-substrate-all-in-one:2024-01-16-pr-2877`

The ledger container image was built with the npm package referenced through the build args, e.g. it has the latest code.

Please re-test everything with these and let me know how it goes (e.g. pass it back for review if you need something from me or if you think everything is working now and it's ready for the final review)

AnmolBansalDEV replied:
@petermetz 
The `connector container`  is giving an unauthorized error to me
logs:
![image](https://github.com/hyperledger/cacti/assets/86472377/4e64afcf-c230-4fe4-a603-2b66061f03b4)
while the `ledger container` is working fine
Please make the image docker image public for me to test it. Thanks!

petermetz replied:
> @petermetz The `connector container` is giving an unauthorized error to me logs: ![image](https://private-user-images.githubusercontent.com/86472377/298798022-4e64afcf-c230-4fe4-a603-2b66061f03b4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5OTgxODMsIm5iZiI6MTcwNTk5Nzg4MywicGF0aCI6Ii84NjQ3MjM3Ny8yOTg3OTgwMjItNGU2NGFmY2YtYzIzMC00ZmU0LWE2MDMtMmI2NjA2MWYwM2I0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIzVDA4MTgwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiNzhmMGZhZTg4YWI2YTA4ZmQ2MDcyN2FlNDgwMjJlYmU0NTM3ODEzNjgxYTAwYWI5MTJlMjljOTJmM2U5MDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yT0obxAhxneoRQMShnZXZexx_xB6LphvwTWm0jVrOxc) while the `ledger container` is working fine Please make the image docker image public for me to test it. Thanks!

@AnmolBansalDEV Oh no, I'm so sorry I forgot to mark it as public on the official repo. Please try again now, it should be visible!

AnmolBansalDEV replied:
@petermetz I've updated the connector with some changes and now it's working. Can you please publish it and then review the PR?

petermetz replied:
> @petermetz I've updated the connector with some changes and now it's working. Can you please publish it and then review the PR?

@AnmolBansalDEV Sure thing, but which one do you mean? The connector container image or the connector npm package?

AnmolBansalDEV replied:
@petermetz I meant the npm package, but I think we'll still have to rebuilt the connector image with the new npm package as the input. So all in all you'll have to publish a new npm package version and then rebuilt the docker image of the connector and then publish it to GitHub docker registery.

petermetz replied:
@AnmolBansalDEV Gotcha, I just did both, you c an find the container image here: `ghcr.io/hyperledger/cactus-connector-polkadot:2024-02-07--2.0.0-lfx-connector.208--pr-2877`

AnmolBansalDEV replied:
Thank you @petermetz

petermetz replied:
> Thank you @petermetz

@AnmolBansalDEV My pleasure and thank you again for the contribution!

petermetz reviewed (changes requested):
@AnmolBansalDEV Please 
1. Update the yarn.lock file (I've just did a rebase locally and then `yarn install` made changes to the lock file so you need to do the same and then commit the lock file)
2. Rename `dockerfile` to `Dockerfile` just to match the conventions.

petermetz commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts:
```diff
@@ -0,0 +1,179 @@
+    await plugin.createAPI();
+
+    const listenOptions: IListenOptions = {
+      hostname: "0.0.0.0",
```
@AnmolBansalDEV Please bind to `localhost` instead otherwise the tests will fail on some cloud provider's instances where the networking is locked down in a way that you can't bind to all interfaces by design.

outSH replied:
`127.0.0.1` or the tests will fail in node 18+ too :)

AnmolBansalDEV replied:
updated
https://github.com/hyperledger/cacti/blob/87ec994040fade0e7a4a690883c971baa9eb9702/packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts#L96-L100

petermetz commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts:
```diff
@@ -0,0 +1,179 @@
+    await plugin.getOrCreateWebServices();
+  });
+  beforeAll(async () => {
+    const rawWasm = fs.readFileSync(
```
@AnmolBansalDEV Please use the async versions of the file reading so that it's not blocking. I recommend the `fs-extra` package which is a great library that has handy async versions of the `fs` package

AnmolBansalDEV replied:
switched to `fs-extra` package's readFile fn
https://github.com/hyperledger/cacti/blob/87ec994040fade0e7a4a690883c971baa9eb9702/packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts#L110-L111

petermetz commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts:
```diff
@@ -0,0 +1,179 @@
+      methodName: "flip",
+      accountAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY", //Alice account address
+      web3SigningCredential: {
+        type: "CACTUS_KEYCHAIN_REF",
```
@AnmolBansalDEV Please make sure to define constants like this in the openapi.json file which will then also make it available for import via the generated typescript code and then you can use those code constants instead of magic strings.

AnmolBansalDEV replied:
updated all the constants to openapi.json's code constants

petermetz commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts:
```diff
@@ -0,0 +1,179 @@
+  });
+  test("flip() invocation", async () => {
+    const result = await apiClient.invokeContract({
+      invocationType: "SEND",
```
@AnmolBansalDEV Please make sure to define constants like this in the openapi.json file which will then also make it available for import via the generated typescript code and then you can use those code constants instead of magic strings.

AnmolBansalDEV replied:
updated all the constants to openapi.json's code constants

petermetz commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts:
```diff
@@ -0,0 +1,179 @@
+  });
+  test("query ink! contract", async () => {
+    const result = await apiClient.invokeContract({
+      invocationType: "QUERY",
```
@AnmolBansalDEV Please make sure to define constants like this in the openapi.json file which will then also make it available for import via the generated typescript code and then you can use those code constants instead of magic strings.

AnmolBansalDEV replied:
updated all the constants to openapi.json's code constants

petermetz commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts:
```diff
@@ -0,0 +1,179 @@
+      methodName: "invalid",
+      accountAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY", //Alice account address
+      web3SigningCredential: {
+        type: "CACTUS_KEYCHAIN_REF",
```
@AnmolBansalDEV Please make sure to define constants like this in the openapi.json file which will then also make it available for import via the generated typescript code and then you can use those code constants instead of magic strings.

AnmolBansalDEV replied:
updated all the constants to openapi.json's code constants

petermetz commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts:
```diff
@@ -0,0 +1,179 @@
+      gasLimit: gasLimit,
+      storageDepositLimit: null,
+      salt: new Uint8Array().toString(),
+      web3SigningCredential: { type: "MNEMONIC_STRING", mnemonic: "//Alice" },
```
@AnmolBansalDEV Please make sure to define constants like this in the openapi.json file which will then also make it available for import via the generated typescript code and then you can use those code constants instead of magic strings.

AnmolBansalDEV replied:
updated
https://github.com/hyperledger/cacti/blob/87ec994040fade0e7a4a690883c971baa9eb9702/packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts#L117-L122

petermetz commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts:
```diff
@@ -0,0 +1,179 @@
+  afterAll(async () => await Servers.shutdown(server));
+  beforeAll(async () => {
+    await ledger.start();
+    expect(ledger).toBeTruthy();
```
@AnmolBansalDEV It would be more useful to check for the output of the `ledger.start()` to be checked for truthyness because the program will 100% crash before this assertion if the `ledger` object was not truthy (because we already tried to call a method on it in the previous line)

AnmolBansalDEV replied:
updated
https://github.com/hyperledger/cacti/blob/87ec994040fade0e7a4a690883c971baa9eb9702/packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts#L74-L75

petermetz reviewed (changes requested):
@AnmolBansalDEV FYI: I left a few comments on one of the test cases but the same change requests also apply for all the other test cases as well

petermetz reviewed (changes requested):
@AnmolBansalDEV FYI: I left a few comments on one of the test cases but the same change requests also apply for all the other test cases as well

AnmolBansalDEV commented:
@petermetz 
I've rebased the PR onto the main branch and incorporated the changes mentioned.

petermetz reviewed (changes requested):
@AnmolBansalDEV

I published the current branch's state as this canary release that you can use for the container:
`@hyperledger/cactus-plugin-ledger-connector-polkadot@2.0.0-lfx-connector.163`

For my last round of change requests please see the comments above.

In general this is looking good and we should be good to go once these minor changes are also addressed.

cc: @jagpreetsinghsasan

AnmolBansalDEV commented on packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts:
```diff
@@ -0,0 +1,184 @@
+      instanceId: instanceId,
+    };
+    plugin = new PluginLedgerConnectorPolkadot(connectorOptions);
+    await plugin.createAPI();
```
@petermetz I've updated the connector and also the testcases
https://github.com/hyperledger/cacti/blob/f7e7a2a1cd9694327e4d323650bd3fad9b3b80dd/packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/integration/invoke-ink-contract.test.ts#L88-L101

petermetz replied:
@AnmolBansalDEV Awesome, thank you!

AnmolBansalDEV commented:
@petermetz 
I've rebased the PR onto the main branch and incorporated the changes mentioned. Please have a look at my comments.

AnmolBansalDEV commented:
@petermetz please have a look at my response

petermetz commented on .github/workflows/ci.yaml:
```diff
@@ -790,6 +790,31 @@ jobs:
+          restore-keys: |
+            ${{ runner.os }}-yarn-${{ hashFiles('./yarn.lock') }}
+      - run: ./tools/ci.sh
+  cactus-plugin-ledger-connector-polkadot:
+    continue-on-error: false
+    env:
+      FULL_BUILD_DISABLED: true
+      JEST_TEST_PATTERN: packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/(unit|integration|benchmark)/.*/*.test.ts
+      JEST_TEST_RUNNER_DISABLED: false
+      TAPE_TEST_RUNNER_DISABLED: true
+    needs: build-dev
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Use Node.js ${{ env.NODEJS_VERSION }}
+        uses: actions/setup-node@v3.6.0
+        with:
+          node-version: ${{ env.NODEJS_VERSION }}
+      - uses: actions/checkout@v3.5.2
+
```
@AnmolBansalDEV Could you please apply the latest patterns here with path filtering conditional execution? It's some fairly simple boilerplate syntax that you can mostly copy from existing packages. Please see the main branch's ci.yaml file for examples and let me know if there are any questions.

AnmolBansalDEV replied:
@petermetz updated

petermetz replied:
@AnmolBansalDEV Thank you. There's a part of it that you missed (actually using the change detection's output) so I posted a suggestion for it in my review. It's a very simple change but without it the change detection does not work unfortunately so we need it as well.

AnmolBansalDEV replied:
updated

petermetz replied:
> updated

@AnmolBansalDEV Thank you!

RafaelAPB commented:
This is a great job. I appreciate you taking our previous work and complementing it, @AnmolBansalDEV! Please incorporate all of the maintainers' feedback. Almost there!

AnmolBansalDEV commented:
@petermetz I've updated the PR and incorporated the changes.

petermetz commented on .github/workflows/ci.yaml:
```diff
@@ -982,6 +991,31 @@ jobs:
+      JEST_TEST_PATTERN: packages/cactus-plugin-ledger-connector-polkadot/src/test/typescript/(unit|integration|benchmark)/.*/*.test.ts
+      JEST_TEST_RUNNER_DISABLED: false
+      TAPE_TEST_RUNNER_DISABLED: true
+    needs: build-dev
```
```suggestion
    needs:
      - build-dev
      - compute_changed_packages
    if: needs.compute_changed_packages.outputs.plugin-ledger-connector-besu-changed == 'true'
```

AnmolBansalDEV replied:
updated @petermetz

petermetz replied:
@AnmolBansalDEV Awesome, thank you!

petermetz reviewed (changes requested):
> This is a great job. I appreciate you taking our previous work and complementing it, @AnmolBansalDEV! Please incorporate all of the maintainers' feedback. Almost there!

@AnmolBansalDEV Plus one to the above ^^

petermetz reviewed (changes requested):
@AnmolBansalDEV Extra commits have crept in again, please squash as needed. We can also break this PR down into multiple smaller PRs if it makes more sense than squashing all the commits together.

The reason why this is important is  because the CI only runs for each PR once meaning that if you have broken commits as intermediate changes in lets' say commit 2 and 3 but then the 4th commit of the PR fixes those problems then we'll never know that commit 2 and 3 were broken the whole time and they'll get merged to the main branch anyway.

AnmolBansalDEV commented:
@petermetz But the PR has 4 progressive commits as mentioned earlier in the PR description as well, should I squash the 4 commits into 1?

petermetz commented:
> @petermetz But the PR has 4 progressive commits as mentioned earlier in the PR description as well, should I squash the 4 commits into 1?

I'd open separate PRs - one PR per commit. So you'd need 3 more. I'd mark  each PR up with the `Depends on ` syntax for the bot that we use: https://github.com/marketplace/actions/dependent-issues

That way a dependency chain can be constructed where the order of the PR's getting merged will be the order  you want them to be.

outSH commented:
@AnmolBansalDEV Please squash all commits or open separate PRs as suggested by peter so we can continue with merging :)

AnmolBansalDEV commented:
@petermetz @outSH I've squashed the commits into one. Looking forward to seeing this PR get merged :)

petermetz reviewed (approved):
@AnmolBansalDEV Thank you very much!!!

petermetz reviewed (changes requested):
@AnmolBansalDEV Well, one last push to fix the checks. Please rebase onto upstream main and fix the codegen action issues:

```sh
Run actions/github-script@v6.4.1
Error: yarn codegen script produced version control side-effects: source files have been changed by it that are otherwise are under version control. This means (99% of the time) that you need to run the yarn codegen script locally and then include the changes it makes in your own commit when submitting your pull request.
```

petermetz reviewed (approved):
@AnmolBansalDEV I re-generated the yarn lock file and updated the commit message and the PR title. LGTM now

petermetz commented:
@AnmolBansalDEV Also fixed the linter errors and the spell checker issues.

AnmolBansalDEV commented:
@petermetz thank you so much for fixing the issues! Is it now ready to be merged?

petermetz commented:
I believe so, but let's see if all the required tests are passing or not. I
configured it so that if they do pass then it gets merged automatically.

If some fail we'll just fix them, no problem.

On Mon, Feb 26, 2024, 9:41 PM Anmol Bansal ***@***.***> wrote:

> @petermetz <https://github.com/petermetz> thank you so much for fixing
> the issues! Is it now ready to be merged?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperledger/cacti/pull/2877#issuecomment-1965827314>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMCQ76W5GQPEAGPY255KF3YVVWXXAVCNFSM6AAAAAA7HP73IOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRVHAZDOMZRGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

petermetz commented:
@AnmolBansalDEV I also had to fix the `codegen` script failing because of some updates to the way we do the openapi generator ignore file handling nowadays. This was the last required test that was failing as far as I could tell so it should actually get merged very soon.

I also updated the commit message to include the issue linking declarations.

AnmolBansalDEV commented:
@petermetz the codegen test is still failing, I don't understand what's causing this. Can you please point me out how to fix this? I tried running `yarn codegen` locally but it gets stuck I guess due to the limitations of my EC2 instance. Can you please look into it?

petermetz commented:
> @petermetz the codegen test is still failing, I don't understand what's causing this. Can you please point me out how to fix this? I tried running `yarn codegen` locally but it gets stuck I guess due to the limitations of my EC2 instance. Can you please look into it?

@AnmolBansalDEV Oops, I forgot to delete a local ignore file. Should be good now, let's see.

BTW: What are the hardware specs on the EC2 machine that you are using? I'd want to know especially the amount of RAM It has.

petermetz commented:
@AnmolBansalDEV Congrats! Wohoo!

AnmolBansalDEV commented:
> > @petermetz the codegen test is still failing, I don't understand what's causing this. Can you please point me out how to fix this? I tried running `yarn codegen` locally but it gets stuck I guess due to the limitations of my EC2 instance. Can you please look into it?
> 
> @AnmolBansalDEV Oops, I forgot to delete a local ignore file. Should be good now, let's see.
> 
> BTW: What are the hardware specs on the EC2 machine that you are using? I'd want to know especially the amount of RAM It has.

oh okay, I think the EC2 instance have 16GB RAM

AnmolBansalDEV commented:
@petermetz Thank you so much, it feels like achieving a major milestone! 
Couldn't have been possible without the continuous support from @jagpreetsinghsasan 

Also thanks to @outSH for valuable code reviews, and to @RafaelAPB and @CatarinaPedreira for foundational work!

RafaelAPB commented:
@AnmolBansalDEV it is a major milestone :D Congratulations for a great connector and thanks for your hard work!
